### ISSUE TITLE
ArrayList payloads lose their elements

### ISSUE THREAD

**Issue by carolmorneau** [spring-cloud/spring-cloud-function#1112]
**Describe the issue**
Message payloads of type `ArrayList` lose their elements as they go through the Spring Cloud Stream framework.  

For instance, an application configured with:
```
spring:
  cloud:
    function:
      definition: myFunction
    stream:
      default:
        producer:
          use-native-encoding: true
```
and a function defined as:
```
	@Bean
	public Function<Message<?>, Message<?>> myFunction() {
		return msg -> msg;
	}
```
the output destination receives the Message with an ArrayList payload, however, all elements from the ArrayList have vanished.  

It appears that this issue was introduced as part of [the following changes in Spring Cloud Function](https://github.com/spring-cloud/spring-cloud-function/commit/8e9948c4c3fc18824408c263938e70f8e416545e). When I undo the changes from that PR, the issue goes away. 

**To Reproduce**
I have uploaded a [minimal application reproducing the issue here](https://github.com/carolmorneau/scst-arraylist-payload-lost). 
Run the unit test called `testSpringCloudFunctionIssueWhereArrayListPayloadsLoseTheirElements` to reproduce the issue. 

**Version of the framework**
Using version `4.1.0` of Spring Cloud Stream and Spring Cloud Function
```
[INFO] +- org.springframework.cloud:spring-cloud-stream:jar:4.1.0:compile
[ ... TRUNCATED ]
[INFO] |  +- org.springframework.cloud:spring-cloud-function-context:jar:4.1.0:compile
[INFO] |  |  +- net.jodah:typetools:jar:0.6.2:compile
[INFO] |  |  +- org.springframework.boot:spring-boot-autoconfigure:jar:3.2.2:compile
[INFO] |  |  +- org.springframework.cloud:spring-cloud-function-core:jar:4.1.0:compile
```
**Expected behavior**
With the above function, I would expect to receive the message with complete arrayList payload on the output destination.

**Comment by carolmorneau** [comment]
FYI, since this issue was opened, we found two workarounds in order to avoid loosing payload elements: 

- Configure consumer's `user-native-decoding` and producer's `use-native-encoding` to `true`. [This currently only works when function composition isn't used](https://github.com/spring-cloud/spring-cloud-function/issues/1144). This essentially does not attempt payload conversion and avoids the issue.
- Provide a [user-defined message converter](https://docs.spring.io/spring-cloud-stream/docs/current/reference/html/spring-cloud-stream.html#spring-cloud-stream-overview-user-defined-message-converters) capable of handling payloads of type `ArrayList`. This ensures that at least one MessageConverter can handle ArrayList payloads and avoids loosing payloads. 

Also, there is a new sample app based on the [function-composition-rabbit](https://github.com/spring-cloud/spring-cloud-stream-samples/tree/main/multi-functions-samples/function-composition-rabbit) sample to showcase the issue. 

Reproduction steps:

* Run the sample app with the issue1112 profile
* Send some data to the RabbitMQ exchange (at least 3 messages since batch-size=3)
* Notice that payloads elements are lost

As a side note, for our specific use-case, https://github.com/spring-cloud/spring-cloud-function/issues/1144 is deemed more important than this issue.

**Comment by olegz** [comment]
I believe this issue has been addressed with https://github.com/spring-cloud/spring-cloud-function/issues/1145, but i will recheck

**Comment by olegz** [comment]
I confirm on my end that it was fixed with https://github.com/spring-cloud/spring-cloud-function/issues/1145, but awaiting for your confirmation

**Comment by carolmorneau** [comment]
The fix on `4.1.2-SNAPSHOT` looks good to me. 

Thank you @olegz for the fix
